README.md
# Spotify Karaoke

[![Website](https://img.shields.io/website?down_color=lightgray&down_message=offline&style=flat-square&up_color=green&up_message=online&url=https%3A%2F%2Fspotify-karaoke.netlify.app)](spotify-karaoke.netlify.app/)
[![Netlify Status](https://img.shields.io/netlify/39fabbc6-dcc1-40e8-8702-a3b798a130e5?style=flat-square)](spotify-karaoke.netlify.app/)
[![Top Lang](https://img.shields.io/github/languages/top/ArturBa/spotify-karaoke?style=flat-square)](https://github.com/ArturBa/spotify-karaoke/search?l=typescript)
[![Licence](https://img.shields.io/github/license/ArturBa/spotify-karaoke?style=flat-square)](./LICENCE)
[![Release](https://img.shields.io/github/v/release/ArturBa/spotify-karaoke?style=flat-square)](https://github.com/ArturBa/spotify-karaoke/releases)
[![Build and Test](https://img.shields.io/github/workflow/status/ArturBa/spotify-karaoke/Build%20and%20test?label=Build%20and%20Test&style=flat-square)](https://github.com/ArturBa/spotify-karaoke/actions/workflows/build_test.yml)
[![Dependabot](https://flat.badgen.net/github/dependabot/ArturBa/spotify-karaoke)](https://github.com/apps/dependabot)

## Why it's created

I wanted to try doing something bigger in Angular framework.

I like to sing along songs and spotify doesn't have a lyrics.

I've tried to implement that for them.

## Preview

![Preview gif](./docs/gif/preview.gif)

## Used resources

- [minilyrics proxy](https://github.com/olee/minilyrics-proxy). Used to get a lyrics from a MiniLyrics. Thanks for your work
- [angular spotify](https://github.com/trungk18/angular-spotify). You need to check that out. This was a big inspiration for my project
- [lcr parser](https://github.com/anhthii/lrc-parser). I've adopted this parser for a getting a lyrics in object form

## Structure

```bash
root
├── apps
│   └── angular-spotify
└── libs
    ├── shared
    │   ├── interceptors (jest)
    │   ├── service (test)
    │   ├── test-helpers
    │   └── view (jest)
    └── web
        ├── lyrics
        │   ├── mini-lyrics
        │   ├── model (jest)
        │   └── view (jest)
        └── spotify
            ├── homepage
            ├── player
            ├── shared
            │   ├── helper (jest)
            │   ├── pipe (jest)
            │   ├── service
            │   └── view (Storybook)
            └── sidebar
```

## More details

Read more about:

- [Internationalization](./docs/i18n.md)
- [Technologies](./docs/technologies.md)
- [Tests](./docs/tests.md)
- [CI/CD](./docs/ci_cd.md)


CHANGELOG.md
# [0.9.0](https://github.com/ArturBa/spotify-karaoke/compare/v0.8.0...v0.9.0) (2021-10-19)

### Features

- **shared:** :sparkles: Add error snackbar ([#205](https://github.com/ArturBa/spotify-karaoke/issues/205)) ([e29d408](https://github.com/ArturBa/spotify-karaoke/commit/e29d4082685b090346e2ff8b3457c8810702e7df))
- **shared:** :sparkles: Add language selection ([#247](https://github.com/ArturBa/spotify-karaoke/issues/247)) ([7f153f5](https://github.com/ArturBa/spotify-karaoke/commit/7f153f58111a6a61cc99aeff469e10fe058259b2))
- **spotify:** :sparkles: Add smartphone icon for user devices ([#219](https://github.com/ArturBa/spotify-karaoke/issues/219)) ([f56fbbc](https://github.com/ArturBa/spotify-karaoke/commit/f56fbbcf9b6470365514ad892290a2521594f9cd))

### Performance Improvements

- :zap: Migrate to yarn ([#217](https://github.com/ArturBa/spotify-karaoke/issues/217)) ([2ef3ecd](https://github.com/ArturBa/spotify-karaoke/commit/2ef3ecd0f75b6ad8139a19696c5e4d0459c4e164))
- **shared:** :lock: Add more strict rules for cookies ([#239](https://github.com/ArturBa/spotify-karaoke/issues/239)) ([6ba15d7](https://github.com/ArturBa/spotify-karaoke/commit/6ba15d726f1e9bafb5a07ff3b232a284f797c905))

# [0.8.0](https://github.com/ArturBa/spotify-karaoke/compare/v0.7.0...v0.8.0) (2021-08-16)

### Bug Fixes

- **lyrics:** :bug: Split title to not use '- From...' in lyrics search ([#202](https://github.com/ArturBa/spotify-karaoke/issues/202)) ([4f66956](https://github.com/ArturBa/spotify-karaoke/commit/4f66956a112234d2825b63ed3dfada0a781d0f07))
- **spotify:** :bug: Use the grid on the player to prevent for moving a track controls ([fb0d26c](https://github.com/ArturBa/spotify-karaoke/commit/fb0d26ce9f1ee28af6950ec43a181b0e32ee29b5))

### Features

- **shared:** :sparkles: Add auth token autorefresh ([#213](https://github.com/ArturBa/spotify-karaoke/issues/213)) ([b2073bb](https://github.com/ArturBa/spotify-karaoke/commit/b2073bb3b08d8a90e0261013ab9097c31c699b49))
- **shared:** :zap: Use proxy for getting user location ([#210](https://github.com/ArturBa/spotify-karaoke/issues/210)) ([e7d6ee1](https://github.com/ArturBa/spotify-karaoke/commit/e7d6ee140e8ae9622b70c6100e07dd1c618c7827))
- **spotify:** :sparkles: Add non premium user message ([#204](https://github.com/ArturBa/spotify-karaoke/issues/204)) ([98385bb](https://github.com/ArturBa/spotify-karaoke/commit/98385bbcb567a3c81e2348236b0a9863234f86cb))
- **spotify:** :sparkles: Add overflow directive ([#203](https://github.com/ArturBa/spotify-karaoke/issues/203)) ([891c3a1](https://github.com/ArturBa/spotify-karaoke/commit/891c3a10f41942b3e8b723b12f6f41d65cf5e19f))
- **spotify:** :sparkles: Add playlists on homepage ([#201](https://github.com/ArturBa/spotify-karaoke/issues/201)) ([cfd4acf](https://github.com/ArturBa/spotify-karaoke/commit/cfd4acf1e25c3b1f9f43b1fc1f50edc968a939c5))
- **spotify:** :sparkles: Add recently played and featured playlists on dashboard ([#215](https://github.com/ArturBa/spotify-karaoke/issues/215)) ([57e63c0](https://github.com/ArturBa/spotify-karaoke/commit/57e63c02dbb5006edb2a4d81afe5085e661cb827))

### Performance Improvements

- :zap: Split homepage views into modules for better first load impression ([#211](https://github.com/ArturBa/spotify-karaoke/issues/211)) ([357e414](https://github.com/ArturBa/spotify-karaoke/commit/357e414e7e478c0b32dbfb4937885fe5c2f6462b))
- **shared:** :chart_with_upwards_trend: Improve google analytics tracking ([#214](https://github.com/ArturBa/spotify-karaoke/issues/214)) ([5493618](https://github.com/ArturBa/spotify-karaoke/commit/54936185fcf22d4b9d0f9f8af7993e68696948f5))

# [0.7.0](https://github.com/ArturBa/spotify-karaoke/compare/v0.6.0...v0.7.0) (2021-08-14)

### Features

- :lipstick: Add homepage padding on large screens ([#196](https://github.com/ArturBa/spotify-karaoke/issues/196)) ([9177aae](https://github.com/ArturBa/spotify-karaoke/commit/9177aae32cdeff47fdf7f84d9e5d6f20f5efed53))
- **lyrics:** :sparkles: Filter lyrics to use only lrc files ([#194](https://github.com/ArturBa/spotify-karaoke/issues/194)) ([55737a5](https://github.com/ArturBa/spotify-karaoke/commit/55737a56364a2a16203ef8abb8fb91e7431091b8))
- **spotify:** :sparkles: Add play button on track ([#197](https://github.com/ArturBa/spotify-karaoke/issues/197)) ([f5cac99](https://github.com/ArturBa/spotify-karaoke/commit/f5cac99f650199191b86208f5abf8eaa56c9c8a8))
- **spotify:** :sparkles: Add play on details view ([#195](https://github.com/ArturBa/spotify-karaoke/issues/195)) ([c1a7372](https://github.com/ArturBa/spotify-karaoke/commit/c1a737280adc698bd53dd63e0e798e36426b20ce))

# [0.6.0](https://github.com/ArturBa/spotify-karaoke/compare/v0.5.0...v0.6.0) (2021-08-12)

### Bug Fixes

- Add a login authorization info ([#159](https://github.com/ArturBa/spotify-karaoke/issues/159)) ([030ab72](https://github.com/ArturBa/spotify-karaoke/commit/030ab72dbbce6af59584dcd13d5d6125ae09538d))
- **components:** :lipstick: Card wrapper single line height ([#174](https://github.com/ArturBa/spotify-karaoke/issues/174)) ([162ed6d](https://github.com/ArturBa/spotify-karaoke/commit/162ed6da53486e97881a65435874eb9046498c04))
- **interceptor:** :bug: Fix refresh token login in interceptor ([#175](https://github.com/ArturBa/spotify-karaoke/issues/175)) ([185446a](https://github.com/ArturBa/spotify-karaoke/commit/185446ac8166ec3a1f405fdbd1ce16be1b96d725))
- **shared:** :bug: Remove duplicate user country requests ([#178](https://github.com/ArturBa/spotify-karaoke/issues/178)) ([35bedfe](https://github.com/ArturBa/spotify-karaoke/commit/35bedfef8705eef9ac61447d41a0db141a581916))

### Features

- **spotify:** :sparkles: Add a device selection ([#168](https://github.com/ArturBa/spotify-karaoke/issues/168)) ([5159c07](https://github.com/ArturBa/spotify-karaoke/commit/5159c078c95195dcbac76244e16935e54ebf347f))

minilyricsApi.ts
import http from 'http';
import crypto from 'crypto';

export interface MinilyricsResponse {
  _type: 'return';
  orgcmd: string;
  result: 'OK' | 'NOT_FOUND';
  badrc: string;
  ls_dd: string;
  server_url: string;
  children: LyricsItem[];
}

export interface LyricsItem {
  _type: 'fileinfo';
  link: string;
  artist?: string;
  title?: string;
  album?: string;
  uploader?: string;
  rate?: string;
  ratecount?: number;
  downloads?: number;
  timelength?: number;
}

export default function queryMiniLyrics(
  title: string,
  artist: string,
  cb: (error: Error | undefined, data?: MinilyricsResponse) => void
) {
  // Send request
  const req = http.request(
    {
      hostname: 'search.crintsoft.com',
      path: '/searchlyrics.htm',
      method: 'POST',
      headers: {
        'User-Agent': 'MiniLyrics',
      },
    },
    (response) => {
      let buffer: Buffer | undefined;
      response.on('data', (buff) => {
        buffer = !buffer ? buff : Buffer.concat([buffer, buff]);
      });
      response.on('end', function () {
        const reader = new CompressedXmlReader(decryptBuffer(buffer!));
        const result = reader.read() as MinilyricsResponse;
        cb(undefined, result);
      });
    }
  );
  req.on('error', (error) => {
    cb(error);
  });

  const query =
    `<?xml version='1.0' encoding='utf-8' ?>` +
    `<searchV1 filetype="lyrics" ClientCharEncoding="utf-8" artist="${
      artist || ''
    }" title="${title || ''}" client="MiniLyrics" />`;
  const reqBuffer = Buffer.from(encryptString(query));

  req.write(reqBuffer);
  req.end();
}

function strToUtf8Bytes(str: string) {
  const utf8: number[] = [];
  for (let ii = 0; ii < str.length; ii++) {
    let charCode = str.charCodeAt(ii);
    if (charCode < 0x80) utf8.push(charCode);
    else if (charCode < 0x800) {
      utf8.push(0xc0 | (charCode >> 6), 0x80 | (charCode & 0x3f));
    } else if (charCode < 0xd800 || charCode >= 0xe000) {
      utf8.push(
        0xe0 | (charCode >> 12),
        0x80 | ((charCode >> 6) & 0x3f),
        0x80 | (charCode & 0x3f)
      );
    } else {
      ii++;
      // Surrogate pair:
      // UTF-16 encodes 0x10000-0x10FFFF by subtracting 0x10000 and
      // splitting the 20 bits of 0x0-0xFFFFF into two halves
      charCode =
        0x10000 + (((charCode & 0x3ff) << 10) | (str.charCodeAt(ii) & 0x3ff));
      utf8.push(
        0xf0 | (charCode >> 18),
        0x80 | ((charCode >> 12) & 0x3f),
        0x80 | ((charCode >> 6) & 0x3f),
        0x80 | (charCode & 0x3f)
      );
    }
  }
  return utf8;
}

function utf8ArrayToString(aBytes: number[] | Uint8Array | Buffer) {
  var sView = '';
  for (var nPart, nLen = aBytes.length, nIdx = 0; nIdx < nLen; nIdx++) {
    nPart = aBytes[nIdx];
    sView += String.fromCharCode(
      nPart > 251 && nPart < 254 && nIdx + 5 < nLen /* six bytes */
        ? /* (nPart - 252 << 30) may be not so safe in ECMAScript! So...: */
          (nPart - 252) * 1073741824 +
            ((aBytes[++nIdx] - 128) << 24) +
            ((aBytes[++nIdx] - 128) << 18) +
            ((aBytes[++nIdx] - 128) << 12) +
            ((aBytes[++nIdx] - 128) << 6) +
            aBytes[++nIdx] -
            128
        : nPart > 247 && nPart < 252 && nIdx + 4 < nLen /* five bytes */
        ? ((nPart - 248) << 24) +
          ((aBytes[++nIdx] - 128) << 18) +
          ((aBytes[++nIdx] - 128) << 12) +
          ((aBytes[++nIdx] - 128) << 6) +
          aBytes[++nIdx] -
          128
        : nPart > 239 && nPart < 248 && nIdx + 3 < nLen /* four bytes */
        ? ((nPart - 240) << 18) +
          ((aBytes[++nIdx] - 128) << 12) +
          ((aBytes[++nIdx] - 128) << 6) +
          aBytes[++nIdx] -
          128
        : nPart > 223 && nPart < 240 && nIdx + 2 < nLen /* three bytes */
        ? ((nPart - 224) << 12) +
          ((aBytes[++nIdx] - 128) << 6) +
          aBytes[++nIdx] -
          128
        : nPart > 191 && nPart < 224 && nIdx + 1 < nLen /* two bytes */
        ? ((nPart - 192) << 6) + aBytes[++nIdx] - 128 /* nPart < 127 ? */
        : /* one byte */
          nPart
    );
  }
  return sView;
}

function encryptString(query: string) {
  const queryBytes = strToUtf8Bytes(query);
  const md5Bytes = crypto
    .createHash('md5')
    .update(query + 'Mlv1clt4.0')
    .digest();

  // Generate encryption key
  const byteSum = queryBytes.reduce((v, x) => v + x, 0);
  const key = Math.floor(byteSum / queryBytes.length) & 255;

  // Value is encrypted
  for (let i = 0; i < queryBytes.length; i++) {
    queryBytes[i] = key ^ queryBytes[i];
  }

  // Final data byte array
  let data = [0x02, key, 0x04, 0x00, 0x00, 0x00, ...md5Bytes, ...queryBytes];
  return data;
}

function decryptBuffer(byteBuffer: Buffer) {
  const k = byteBuffer.readInt8(1) & 255;
  const valueBuffer = byteBuffer.slice(22, byteBuffer.length);
  for (let i = 0; i < valueBuffer.length; i++) {
    valueBuffer[i] = (valueBuffer.readInt8(i) ^ k) & 255;
  }
  return valueBuffer;
}

class CompressedXmlReader {
  public position = 0;

  private stringTable: string[] = [];

  constructor(public readonly buffer: Buffer) {}

  private peekByte() {
    return this.buffer.readUInt8(this.position);
  }

  private readByte() {
    const v = this.peekByte();
    this.position++;
    return v;
  }

  private expectByte(x: number) {
    const v = this.readByte();
    if (v !== x) {
      throw new Error(`Expected (${x}), but got (${v})`);
    }
    return v;
  }

  private readChar() {
    return String.fromCharCode(this.readByte());
  }

  private readInt() {
    const v = this.buffer.readUInt32LE(this.position);
    this.position += 4;
    return v;
  }

  private readSlice(byteCount?: number) {
    if (byteCount === undefined)
      byteCount = this.buffer.length - this.position + 1;
    const v = this.buffer.slice(this.position, this.position + byteCount);
    this.position += byteCount;
    return v;
  }

  private popValue() {
    return this.stringTable[this.readByte() - 10];
  }

  private readHeader() {
    if (
      this.readChar() !== 'M' ||
      this.readChar() !== 'B' ||
      this.readChar() !== 'X' ||
      this.readChar() !== 'M' ||
      this.readChar() !== 'L' ||
      this.readChar() !== '1'
    ) {
      throw new Error('MBXML header missmatch');
    }
    if (this.readInt() !== 2) throw new Error('Header version mismatch');
    // end of stream is encoded here, but it counts starting from the header (MBXML1)
    const fileEnd = this.readInt() - 6 - 4 - 4 - 2;
  }

  private readStringTable() {
    if (this.readChar() !== 'S' || this.readChar() !== 'T') {
      throw new Error('String table header missmatch');
    }

    // size of string table in bytes
    // (including both int32 values at the beginning)
    const stSize = this.readInt() - 8;

    // Number of strings in the string table
    const stCount = this.readInt();

    // console.log('fileEnd =', fileEnd, fileEnd.toString(16));
    // console.log('stSize =', stSize, stSize.toString(16));
    // console.log('stCount =', stCount);
    // Convert utf-8 bytes in string table to string

    const stringTableStr = utf8ArrayToString(this.readSlice(stSize));

    // Split by 0-byte
    this.stringTable = stringTableStr.split('\x00');
    this.stringTable.pop();

    if (this.stringTable.length !== stCount) {
      console.warn(
        `String table should contain ${stCount} items, but has ${this.stringTable.length}`
      );
    }
  }

  private readElement() {
    this.expectByte(2);
    const node: any = { _type: this.popValue() };
    // Parse attributes
    while (this.peekByte() >= 10) {
      const key = this.stringTable[this.readByte() - 10];
      const value = this.stringTable[this.readByte() - 10];
      node[key] = value;
    }
    // Parse children if present
    if (this.peekByte() === 3) {
      this.position++;
      node.children = [];
      // loop until no next child found
      while (this.peekByte() !== 4) {
        node.children.push(this.readElement());
      }
    }
    this.expectByte(4);
    return node;
  }

  public read() {
    this.readHeader();

    // String table (ST) start
    this.readStringTable();

    // Number of xml entries
    const nodeCount = this.readInt();
    return this.readElement();
  }
}
